libsolutil/IpfsHash.cpp

```diff
@@ -55,34 +55,125 @@ string base58Encode(bytes const& _data)
+///   - Do this until the current level has only one node, return the hash in that node
+bytes groupChunksBottomUp(vector<Chunks>& chunkLevels)
+{
+	size_t const maxChildNum = 174;
+	size_t currentLevel = 0;
+
+	// when we reach root it will be the only node in that level
+	while (chunkLevels[currentLevel].size() != 1)
+	{
+		vector<Chunk> nextLevel;
+		Chunk newChunk;
+		bytes newLinks;
+		bytes newLengths;
+		size_t newChunkSize = 0;
+		size_t newBlockSize = 0;
+
+		for (size_t chunkIndex = 0; chunkIndex < chunkLevels[currentLevel].size(); chunkIndex++)
+		{
+			// new link that we are adding
+			bytes newLink = bytes {0x0a} + varintEncoding(chunkLevels[currentLevel][chunkIndex].hash.size());
+			newLink += std::move(chunkLevels[currentLevel][chunkIndex].hash);
+			newLink += bytes{0x12, 0x00, 0x18} + varintEncoding(chunkLevels[currentLevel][chunkIndex].blockSize);
+			newLinks += bytes{0x12} + varintEncoding(newLink.size()) + std::move(newLink);
+
+			newLengths += bytes{0x20} + varintEncoding(chunkLevels[currentLevel][chunkIndex].size);
+
+			newChunkSize += chunkLevels[currentLevel][chunkIndex].size;
+			newBlockSize += chunkLevels[currentLevel][chunkIndex].blockSize;
+
+			if (chunkIndex % maxChildNum == maxChildNum - 1 || chunkIndex == chunkLevels[currentLevel].size() - 1)
+			{
+				bytes newData = bytes{0x08, 0x02, 0x18} + varintEncoding(newChunkSize) + newLengths;
+				bytes newLengthsData = bytes{0x0a} + varintEncoding(newData.size()) + std::move(newData);
+
+				bytes blockData = std::move(newLinks) + std::move(newLengthsData);
+
+				nextLevel.emplace_back(
+					bytes{0x12, 0x20} + picosha2::hash256(blockData),
+					newChunkSize,
+					newBlockSize + blockData.size()
+				);
+
+				// we need to reset as we are about to build another node
+				newLinks = {};
+				newLengths = {};
+				newChunkSize = 0;
+				newBlockSize = 0;
+			}
+		}
+		chunkLevels.emplace_back(std::move(nextLevel));
+		chunkLevels[currentLevel].clear();
+		currentLevel++;
+	}
+
+	// top level's only node stores the hash for file
+	return chunkLevels[currentLevel].front().hash;
+}
+
 bytes solidity::util::ipfsHash(string _data)
 {
-	assertThrow(_data.length() < 1024 * 256, DataTooLong, "IPFS hash for large (chunked) files not yet implemented.");
+	size_t const maxChunkSize = 1024 * 256;
+	size_t chunkCount = _data.length() / maxChunkSize + (_data.length() % maxChunkSize > 0 ? 1 : 0);
+	chunkCount = chunkCount == 0 ? 1 : chunkCount;
 
-	bytes lengthAsVarint = varintEncoding(_data.size());
+	vector<Chunks> chunkLevels(1);
 
-	bytes protobufEncodedData;
-	// Type: File
-	protobufEncodedData += bytes{0x08, 0x02};
-	if (!_data.empty())
+	for (unsigned long chunkIndex = 0; chunkIndex < chunkCount; chunkIndex++)
 	{
-		// Data (length delimited bytes)
-		protobufEncodedData += bytes{0x12};
-		protobufEncodedData += lengthAsVarint;
-		protobufEncodedData += asBytes(std::move(_data));
+		std::string chunkData =
```

Comment:
Can you use a `bytesRef`? Then we would not have to copy the string.

Reply:
Or acutally - we use it in `asBytes(...)` below - so maybe it makes sense to convert it to a bytes already here?

Reply:
Totally makes sense to convert to bytes earlier